[feature] Value relation restore missing layers from DBs

elpaso opened the pull request:
Checks for missing layers in value relation widget and loads them automatically (if possible), also warns the user if the layer could not be found.

The API is generic enough to be used by other widgets too, if they happen to require one or more other layers to work correctly.

The use case we address here is when a layer stored in PG or GPKG or SpatiaLite is configured to use a value-relation widget but the required layer was not already loaded in the project (or it was loaded but it was moved to another schema), in that case the weak resolution won't find the layer and the widget will not work. With this implementation, the layer is first searched in the loaded layers (matching the provider name and the layer name) and if it is not found it will be searched in the same connection as the parent layer (first in the same schema, then in the other schemas) and loaded if found.

The implementation also works when applying a style that uses a value relation widget to an existing layer.

Funded by **ARPA Piemonte**

elpaso committed:
Check vector layer widget broken dependencies

elpaso committed:
Load dependency layers automatically when possible

elpaso committed:
Auto load missing layers from Postgis

elpaso committed:
Resolve mssing layers loosely by name and provider

m-kuhn commented on src/app/qgisapp.cpp:
```diff
@@ -1940,6 +1950,137 @@ QgsMessageBar *QgisApp::visibleMessageBar()
+  {
+    std::unique_ptr<QgsEditorWidgetWrapper> ww;
+    ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
+    for ( QgsVectorLayerRef &dependency : ww->layerDependencies() )
```
Take a const copy before iterating a container with a range-based for loop or donald trump riding a unicorn singing elton john songs will appear every night in front of your window. And believe me it sounds terrible when he sings elton john.

elpaso replied:
It's not that I like Elton John, but `resolveWeakly` is not const.

m-kuhn replied:
Hmm... So what happens is you get the `layerDependencies`, take a deep copy of it to iterate, then `resolveWeakly()` on each of the items and drop the modified container again, while forwarding the individual items. This kind of works, so maybe you get the same scenario but muted. If you can live with that, I can as well :+1: I think to get rid of unicorn and president as well you'll need to take a copy only of each item in the loop without a detach in the beginning.

m-kuhn commented on src/app/qgisapp.cpp:
```diff
@@ -6315,6 +6461,18 @@ bool QgisApp::addProject( const QString &projectFile )
 #endif
 
+    // Check for missing layer widget dependencies
+    for ( QgsVectorLayer *vl : QgsProject::instance()->layers<QgsVectorLayer *>( ) )
```
No elton john excuses here ;)

m-kuhn commented on src/app/qgisapp.h:
```diff
@@ -1700,6 +1701,11 @@ class APP_EXPORT QgisApp : public QMainWindow, private Ui::MainWindow
     void onActiveLayerChanged( QgsMapLayer *layer );
 
+    /**
+     * Triggered when a vector layer style has changed, check for widget config layer dependencies
```
```suggestion
     * Emitted when a vector layer style has changed, check for widget config layer dependencies
```

m-kuhn commented:
Some lovely work in here :+1:

elpaso commented on src/app/qgisapp.cpp:
```diff
@@ -1940,6 +1950,137 @@ QgsMessageBar *QgisApp::visibleMessageBar()
+      }
+      if ( ! loaded )
+      {
+        const QString msg { tr( "Form for layer '%1' requires layer '%2' to be loaded but '%2' could not be found, please load it manually if possible." )
```
Yes, for now the broken dependency only comes from a widget configuration, that's why I added "Form", but I'm ok to drop it, I had doubts about it in the first place.

m-kuhn replied:
I just want to be sure we don't forget it when that ever changes, good ideas welcome!

elpaso commented on src/core/qgsvectorlayer.cpp:
```diff
@@ -4687,10 +4688,12 @@ QString QgsVectorLayer::loadNamedStyle( const QString &theURI, bool &resultFlag,
   }
-  return QgsMapLayer::loadNamedStyle( theURI, resultFlag, categories );
+  returnMessage = QgsMapLayer::loadNamedStyle( theURI, resultFlag, categories );
+  emit styleLoaded();
```
good idea!

elpaso committed:
Address PR comments

elpaso committed:
Code layout typos

elpaso committed:
Fix test case and add some more BSTS checks

BSTS=better safe than sorry

nyalldawson commented on src/core/qgsmaplayer.h:
```diff
@@ -1370,6 +1370,13 @@ class CORE_EXPORT QgsMapLayer : public QObject
+     * \param categories style categories
+     * \since QGIS 3.12
+     */
+    void styleLoaded( const QgsMapLayer::StyleCategories &categories );
```
```suggestion
    void styleLoaded( QgsMapLayer::StyleCategories categories );
```

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -660,6 +661,18 @@ void QgisApp::onActiveLayerChanged( QgsMapLayer *layer )
   emit activeLayerChanged( layer );
 }
 
+void QgisApp::vectorLayerStyleLoaded( const QgsMapLayer::StyleCategories &categories )
```
```suggestion
void QgisApp::vectorLayerStyleLoaded( QgsMapLayer::StyleCategories categories )
```

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -660,6 +661,18 @@ void QgisApp::onActiveLayerChanged( QgsMapLayer *layer )
+  if ( categories.testFlag( QgsMapLayer::StyleCategory::Forms ) )
+  {
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( sender() );
+    if ( vl && vl->isValid( ) )
```
Question -- why does this require a valid layer? You could potentially load a style for a layer with a broken path, and still get the dependencies resolved...

elpaso replied:
If the layer is invalid I see no point in resolving dependencies, because you will not able to use them anyways. But, if you fix the broken layer later, then the style is applied again and the resolution will be attempted.

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -1941,6 +1954,143 @@ QgsMessageBar *QgisApp::visibleMessageBar()
+
+void QgisApp::checkVectorLayerDependencies( QgsVectorLayer *vl )
+{
+  if ( vl && vl->isValid() )
```
Again, not sure if we need to restrict this to valid layers

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -1941,6 +1954,143 @@ QgsMessageBar *QgisApp::visibleMessageBar()
+    for ( const QgsVectorLayerRef &dependency : constDependencies )
+    {
+      const QgsVectorLayer *depVl = nullptr;
+      // The default resolution logic does not recognize a positive match if the datasource is
```
This logic should belong in _LayerRef

elpaso replied:
Yeah, I though about it, not sure: do you think at adding another method that `resolveAggresively` ?

But the main reason I wanted it in app and not in core was that I wanted to communicate with the user, what would you suggest? A signal connected to app ?

nyalldawson replied:
I just mean the first part of this branch -- the part where it scans through project layers for ones which match the name and provider

elpaso replied:
Sorry, I don't follow you: `resolveWeakly` does that (but it matches the public source exactly), the only option I see (and that's the one I proposed in my previous comment) is to have a `resolveAggressively` (or whatever you want to name it) that also searches for matching layers by provider AND name. But that won't be sufficient to solve my problem here, because _LayerRef is designed to find layers already defined in the current project while what I'm doing here is searching for layers in the provider that do not exist in the project.

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -5040,6 +5189,12 @@ bool QgisApp::addVectorLayersPrivate( const QStringList &layerQStringList, const
   for ( QgsMapLayer *l : qgis::as_const( layersToAdd ) )
   {
+    // Connect style changed signal to run dependencies checks
+    QgsVectorLayer *vl { qobject_cast< QgsVectorLayer * >( l ) };
```
Instead of implementing this here, I think it would be safer to listen to the project's layer added signals and respond to that. There's many ways to add layers to qgis which don't go through addVectorLayersPrivate

elpaso replied:
Do you mean the connection, right?

elpaso committed:
Update src/app/qgisapp.cpp

Co-Authored-By: Nyall Dawson <nyall.dawson@gmail.com>

elpaso committed:
Update src/core/qgsmaplayer.h

Co-Authored-By: Nyall Dawson <nyall.dawson@gmail.com>

elpaso committed:
Connect vl to vectorLayerStyleLoaded on load

elpaso committed:
Fix build

elpaso committed:
Sipify

elpaso committed:
Remove this from lambda

elpaso committed:
Add match type to resolveWeakly

This way can define exactly how the weak resolution
will consider a layer a postivie match.

elpaso committed:
Changes dependency message: removed 'Form'

m-kuhn commented on src/app/qgisapp.cpp:
```diff
@@ -1941,6 +1954,131 @@ QgsMessageBar *QgisApp::visibleMessageBar()
+  for ( int i = 0; i < vl->fields().count(); i++ )
+  {
+    std::unique_ptr<QgsEditorWidgetWrapper> ww;
+    ww.reset( QgsGui::editorWidgetRegistry()->create( vl, i, nullptr, nullptr ) );
```
@elpaso I get a crash (failing assert) since this change here when loading a project (qgep).

![image](https://user-images.githubusercontent.com/588407/70787927-10a58300-1d90-11ea-872e-3808cfc20cbe.png)

What I wonder now is if the approach to *create* a disposable widget wrapper here is the right way to go. Or if this kind of handling would be better suited in the factory directly (maybe even in the fieldformatter in core since it's not really GUI related and could be used by qfield or theoretically even server too).

elpaso replied:
Do you mean moving `QgsEditorWidgetWrapper::layerDependencies()` logic into the field formatter?

I'm not sure that it's the right place, the configuration about widgets "dependencies" is stored in the widget configuration and it belongs to the widget.

What exactly are you suggesting?

Isn't actually better to check if the assert can be removed putting some safe guards wherever in the widget logic there is a danger of misbehavior due to vectorLayerTools being nullptr?

Or even easier: create the`QgsAttributeEditorContext` and pass it to the factory?

m-kuhn replied:
My preferred approach would indeed be to move it to the field formatter.

The field formatter is also based on the same configuration (see e.g. https://github.com/qgis/QGIS/blob/master/src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp#L119 ) and lives in core so the logic can be used everywhere. They are created once per application lifetime, so there is no need to create disposable objects.

But, since I'm not completely sure that there are no hidden dependencies involved, the second proposal less risky but also less generic would be to move it to the widget factories. They are also only created once per application lifetime.

Your proposal will be the easiest in terms of work (because it follows the structure set up in this PR), but is also mostly putting the dust under the carpet, to resolve layer dependencies there is no widget (gui element) required, so I don't think creating a throw away widget wrapper for this purpose is the best way to implement it.

elpaso replied:
Ok, I'll look into the field formatter approach.

m-kuhn replied:
Thanks.
Let me know if you run into questions!
